### PR TITLE
Repo Delete always updates deleted_at and name

### PR DESCRIPTION
**Pull request by cbart**
Part of #49298

Modify the soft-delete query, to always amend repo name and deleted_at if they are not set to indicate deleted state.

## Test plan

Add tests with database for corner cases.

Check the migration:

```
➜  sourcegraph git:(cb/repo-delete-reconciliation) ✗ sg migration up
✅ Schema(s) are up-to-date!

ℹ️ Auto updating sg ...
⚠️ update check: failed to install update: reading response: context deadline exceeded

➜  sourcegraph git:(cb/repo-delete-reconciliation) ✗ sg migration downto -db=frontend -target=1679404397

ℹ️ Auto updating sg ...
⚠️ update check: failed to install update: reading response: context deadline exceeded

➜  sourcegraph git:(cb/repo-delete-reconciliation) ✗ sg migration up
✅ Schema(s) are up-to-date!

ℹ️ Auto updating sg ...
⚠️ update check: failed to install update: reading response: context deadline exceeded
```

### CODE REVIEW

**Review by sashaostrikov — APPROVED** [review]

File: internal/database/repos.go [hunk 1]
```diff
@@ -1631,11 +1631,14 @@ WITH repo_ids AS (
 )
 UPDATE repo
 SET
-  name = soft_deleted_repository_name(name),
-  deleted_at = transaction_timestamp()
+  name = (CASE
+    WHEN name LIKE 'DELETED%%'
+	THEN name
+	ELSE soft_deleted_repository_name(name)
+  END),
```
Comment (eseliger) [inline internal/database/repos.go:1638]: I wonder if this should be something that soft_deleted_repository_name should always do, vs here inline.

Comment (cbart) [reply internal/database/repos.go:1638]: 🤔 that's a good point.

Comment (cbart) [reply internal/database/repos.go]: :done:

**Review by mrnugget — APPROVED** [review]

File: internal/database/repos_test.go [hunk 2]
```diff
@@ -2497,6 +2497,106 @@ func TestRepos_Delete(t *testing.T) {
...
+	db := NewDB(logger, dbtest.NewDB(logger, t))
+	ctx := context.Background()
+	ctx = actor.WithActor(ctx, &actor.Actor{UID: 1, Internal: true})
+	if err := upsertRepo(ctx, db, InsertRepoOp{Name: "myrepo"}); err != nil {
+		t.Fatal(err)
+	}
+	repo, err := db.Repos().GetByName(ctx, "myrepo")
+	if err != nil {
+		t.Fatal(err)
+	}
```
[the remark this answers is not in the document]
Comment (cbart) [reply internal/database/repos_test.go]: Thanks for the tip!

File: migrations/frontend/squashed.sql [hunk 3]
```diff
@@ -780,10 +780,14 @@ END;
 $$;
 
 CREATE FUNCTION soft_deleted_repository_name(name text) RETURNS text
-    LANGUAGE plpgsql STRICT
+    LANGUAGE plpgsql
     AS $$
 BEGIN
-    RETURN 'DELETED-' || extract(epoch from transaction_timestamp()) || '-' || name;
+    IF name LIKE 'DELETED-%' THEN
```
Comment (indradhanush) [inline migrations/frontend/squashed.sql:786]: Should this be `AND deleted_at is null`? What if a repo itself starts with `DELETED-` as the real name?

Comment (varsanojidan) [reply migrations/frontend/squashed.sql:786]: 👌

**Pull request merged**